rustworkx-core/src/dag_algo.rs

```diff
@@ -333,6 +337,154 @@ where
+                if self.graph.to_index(*node) >= self.graph.node_bound() {
+                    panic!("Node {:#?} is not present in the graph.", node);
+                }
+                if self.cycle_check.contains(node) {
+                    return Some(Err(LayersError(
+                        "An invalid first layer was provided.".to_owned(),
+                    )));
+                }
+                self.cycle_check.insert(*node);
+            }
+            Some(Ok(self.cur_layer.to_owned()))
+        } else if self.cur_layer.is_empty() {
+            None
+        } else {
+            for node in &self.cur_layer {
+                if self.graph.to_index(*node) >= self.graph.node_bound() {
+                    panic!("Node {:#?} is not present in the graph.", node);
+                }
+                let children = self
+                    .graph
+                    .neighbors_directed(*node, petgraph::Direction::Outgoing);
+                let mut used_indices: HashSet<G::NodeId> = HashSet::new();
+                for succ in children {
+                    // Skip duplicate successors
+                    if used_indices.contains(&succ) {
+                        continue;
+                    }
+                    used_indices.insert(succ);
+                    let mut multiplicity: usize = 0;
+                    let raw_edges: G::EdgesDirected = self
+                        .graph
+                        .edges_directed(*node, petgraph::Direction::Outgoing);
+                    for edge in raw_edges {
+                        if edge.target() == succ {
+                            multiplicity += 1;
+                        }
+                    }
+                    self.predecessor_count
+                        .entry(succ)
+                        .and_modify(|e| *e -= multiplicity)
+                        .or_insert(
+                            // Get the number of incoming edges to the successor
+                            self.graph
+                                .edges_directed(succ, petgraph::Direction::Incoming)
+                                .count()
+                                - multiplicity,
+                        );
+                    if *self.predecessor_count.get(&succ).unwrap() == 0 {
+                        if self.cycle_check.contains(&succ) {
+                            return Some(Err(LayersError("The provided graph contains a cycle or an invalid first layer was provided.".to_string())));
+                        }
+                        self.next_layer.push(succ);
+                        self.cycle_check.insert(succ);
+                        self.predecessor_count.remove(&succ);
+                    }
+                }
+            }
+            swap(&mut self.cur_layer, &mut self.next_layer);
+            self.next_layer.clear();
+            if self.cur_layer.is_empty() {
+                None
+            } else {
+                Some(Ok(self.cur_layer.to_owned()))
```

Comment:
I think this a bit more explicit:

```suggestion
                Some(Ok(self.cur_layer.clone()))
```